# 2. 데이터 분석 개요

  앞으로 다섯 시간에 걸쳐 데이터분석에 대해 소개드리겠습니다. 이번 시간은 쥬피터노트북을 통한 실습을 통해 파이썬 기초와 판다스 기초에 대해 가볍게 진행하겠습니다. 온라인으로 한 시간 동안 진행하기에는 분량이 다소 많아 이렇게 문서로 공유드리오니 예습복습에 활용해 주시기 바랍니다.



## 2-1. 일반적인 데이터분석 프로세스

"데이터 분석"이라는 용어는 굉장히 막연한 편입니다. "데이터"도 "분석"도 크게 생소한 용어가 아닌데 말이죠. 어쩌면 "머신러닝"이나 "파이썬", "R"이 유행하면서 뭔가 더 거창한 개념이 되어버린 것 같지만, 본질은 예나 지금이나 동일합니다. ①데이터를 취합하고, ②해당 데이터를 뜯어보면서 **패턴**을 발견하고, ③그림이나 숫자로 요약하는 기술통계를 진행한 후, ④발견한 패턴이 전체 데이터에 적용 가능한지 가설을 세운 후, ⑤유의성을 검정하고, ⑥이를 바탕으로 인사이트를 도출합니다. ⑦이에 기반하여 의사결정을 내리거나 실행에 옮기게 됩니다.

데이터 취합(크롤링, 설문 등)
-> 데이터 전처리(외부 데이터의 가공, 클렌징)를 통한 마스터 데이터셋의 생성
-> 탐색적 데이터 분석(EDA, wrangling)
-> 분석모델 및 알고리즘 결정(분석방법론)
-> 시각화 보고서를 활용한 커뮤니케이션
-> 의사결정 또는 프로덕트 배포

## 2-2. 이번 차시에 알아볼 내용

이번 차시에는 탐색적 데이터 분석(EDA, Exploratory Data Analysis)을 진행해보겠습니다. 파이썬이나 판다스를 통해 커맨드를 통한 분석이 어느 정도만 익숙해지셔도 엑셀보다 훨씬 더 간편하게 데이터를 탐색하고 시각화할 수 있다는 느낌을 갖게 되실 겁니다.

이번 차시에 이용해볼 데이터셋은 "fifa15.csv"입니다. 다운받으셔서 쥬피터노트북 파일(.ipynb, 이하 "노트북")과 동일한 폴더에 넣어주시기 바랍니다.

 # 모듈 임포트

파이썬 모듈을 임포트합니다. 파이썬에서 특정 모듈을 임포트하는 방법은 ```import 모듈명```입니다. 모듈명이 길어서 커맨드를 입력하기 귀찮은 이들을 위해 ```import 모듈명 as 별칭```으로 임포트하기도 합니다. 판다스는 모듈명이 pandas이지만, 보통 pd라는 별칭으로, 넘파이는 numpy지만, 보통 np라는 별칭을 사용합니다. 오래 전부터 굳어진 관례라서 여러분도 특별한 이유가 없다면 이 컨벤션을 따라 주실 것을 권장드립니다. 여러분의 PC에 파이썬을 설치하셨다면 아래의 커맨드를 실행하여(Shift-Enter) 판다스를 설치해 주시기 바랍니다. (아나콘다를 설치하셨거나, 구글콜랩 등을 사용하고 계시다면 모듈이 이미 설치되어 있습니다.)

In [87]:
!pip install pandas

위 명령을 통해 판다스와 넘파이 등이 설치되었습니다. 이제 임포트를 해 보겠습니다. 위에서 말씀드린 대로 각각 ```pd```와 ```np```라는 별칭으로 임포트합니다.

In [88]:
import pandas as pd
import numpy as np

# 데이터셋 불러오기

이제 데이터셋을 불러오겠습니다. 관례상 데이터프레임은 df라는 변수명으로 사용하는 것이 일반적입니다. (데이터프레임을 가공해 가면서 새로운 이름의 데이터프레임을 생성하기도 합니다.)

In [89]:
df = pd.read_csv("fifa21.csv")

이제 df만 입력하고 셀을 실행해보면 ```fifa21.csv```가 데이터프레임 형태로 출력됩니다.

In [90]:
df

sofifa_id                                         player_url  \
0         158023  https://sofifa.com/player/158023/lionel-messi/...   
1          20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
2         190871  https://sofifa.com/player/190871/neymar-da-sil...   
3         167495  https://sofifa.com/player/167495/manuel-neuer/...   
4         176580  https://sofifa.com/player/176580/luis-suarez/2...   
...          ...                                                ...   
18939     256679  https://sofifa.com/player/256679/kevin-angulo/...   
18940     257710  https://sofifa.com/player/257710/mengxuan-zhan...   
18941     250989  https://sofifa.com/player/250989/zhenghao-wang...   
18942     257697  https://sofifa.com/player/257697/zitong-chen/2...   
18943     257936   https://sofifa.com/player/257936/yue-song/210002   

              short_name                            long_name  age  \
0               L. Messi       Lionel Andrés Messi Cuccittini   33   
1      Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro   35   
2              Neymar Jr        Neymar da Silva Santos Júnior   28   
3               M. Neuer                         Manuel Neuer   34   
4              L. Suárez             Luis Alberto Suárez Díaz   33   
...                  ...                                  ...  ...   
18939          K. Angulo                         Kevin Angulo   24   
18940     Zhang Mengxuan                       Mengxuan Zhang   21   
18941      Wang Zhenghao                                  王政豪   20   
18942        Chen Zitong                          Zitong Chen   23   
18943           Song Yue                             Yue Song   28   

              dob  height_cm  weight_kg nationality  \
0      1987-06-24        170         72   Argentina   
1      1985-02-05        187         83    Portugal   
2      1992-02-05        175         68      Brazil   
3      1986-03-27        193         92     Germany   
4      1987-01-24        182         86     Uruguay   
...           ...        ...        ...         ...   
18939  1996-04-13        176         73    Colombia   
18940  1999-04-26        177         70    China PR   
18941  2000-06-28        185         74    China PR   
18942  1997-02-20        186         80    China PR   
18943  1991-11-20        185         79    China PR   

                                 club_name              league_name  \
0                             FC Barcelona   Spain Primera Division   
1                                 Juventus          Italian Serie A   
2                      Paris Saint-Germain           French Ligue 1   
3                        FC Bayern München     German 1. Bundesliga   
4                             FC Barcelona   Spain Primera Division   
...                                    ...                      ...   
18939                      América de Cali  Colombian Liga Postobón   
18940  Chongqing Dangdai Lifan FC SWM Team     Chinese Super League   
18941                      Tianjin TEDA FC     Chinese Super League   
18942        Shijiazhuang Ever Bright F.C.     Chinese Super League   
18943                      Tianjin TEDA FC     Chinese Super League   

       league_rank  overall  potential  value_eur  wage_eur player_positions  \
0              1.0       93         93   67500000    560000       RW, ST, CF   
1              1.0       92         92   46000000    220000           ST, LW   
2              1.0       91         91   90000000    270000          LW, CAM   
3              1.0       89         89   29000000    125000               GK   
4              1.0       87         87   31500000    280000               ST   
...            ...      ...        ...        ...       ...              ...   
18939          1.0       47         52      40000       500               CM   
18940          1.0       47         52      35000      1000               CB   
18941          1.0       47         51      35000      1000               CB   

약 2만 개에 가까운 행의 값을 엑셀처럼 전부 불러와서 확인하는 것은 무의미한 일입니다. 보통은 ```df.head()```와 ```df.tail()``` 메서드를 통해 일부의 값만 확인해봅니다.

In [91]:
df.head(3)  # 괄호 안의 숫자만큼의 행을 출력

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
2     190871  https://sofifa.com/player/190871/neymar-da-sil...   

          short_name                            long_name  age         dob  \
0           L. Messi       Lionel Andrés Messi Cuccittini   33  1987-06-24   
1  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro   35  1985-02-05   
2          Neymar Jr        Neymar da Silva Santos Júnior   28  1992-02-05   

   height_cm  weight_kg nationality            club_name  \
0        170         72   Argentina         FC Barcelona   
1        187         83    Portugal             Juventus   
2        175         68      Brazil  Paris Saint-Germain   

              league_name  league_rank  overall  potential  value_eur  \
0  Spain Primera Division          1.0       93         93   67500000   
1         Italian Serie A          1.0       92         92   46000000   
2          French Ligue 1          1.0       91         91   90000000   

   wage_eur player_positions preferred_foot  international_reputation  \
0    560000       RW, ST, CF           Left                         5   
1    220000           ST, LW          Right                         5   
2    270000          LW, CAM          Right                         5   

   weak_foot  skill_moves    work_rate   body_type real_face  \
0          4            4   Medium/Low       Messi       Yes   
1          4            5     High/Low  C. Ronaldo       Yes   
2          5            5  High/Medium      Neymar       Yes   

   release_clause_eur                                        player_tags  \
0         138400000.0  #Dribbler, #Distance Shooter, #FK Specialist, ...   
1          75900000.0  #Aerial Threat, #Dribbler, #Distance Shooter, ...   
2         166500000.0  #Speedster, #Dribbler, #Playmaker, #FK Special...   

  team_position  team_jersey_number loaned_from      joined  \
0           CAM                10.0         NaN  2004-07-01   
1            LS                 7.0         NaN  2018-07-10   
2            LW                10.0         NaN  2017-08-03   

   contract_valid_until nation_position  nation_jersey_number  pace  shooting  \
0                2021.0              RW                  10.0  85.0      92.0   
1                2022.0              LS                   7.0  89.0      93.0   
2                2022.0             NaN                   NaN  91.0      85.0   

   passing  dribbling  defending  physic  gk_diving  gk_handling  gk_kicking  \
0     91.0       95.0       38.0    65.0        NaN          NaN         NaN   
1     81.0       89.0       35.0    77.0        NaN          NaN         NaN   
2     86.0       94.0       36.0    59.0        NaN          NaN         NaN   

   gk_reflexes  gk_speed  gk_positioning  \
0          NaN       NaN             NaN   
1          NaN       NaN             NaN   
2          NaN       NaN             NaN   

                                       player_traits  attacking_crossing  \
0  Finesse Shot, Long Shot Taker (AI), Speed Drib...                  85   
1  Power Free-Kick, Flair, Long Shot Taker (AI), ...                  84   
2  Injury Prone, Flair, Speed Dribbler (AI), Outs...                  85   

   attacking_finishing  attacking_heading_accuracy  attacking_short_passing  \
0                   95                          70                       91   
1                   95                          90                       82   
2                   87                          62                       87   

   attacking_volleys  skill_dribbling  skill_curve  skill_fk_accuracy  \
0                 88               96           93                 94   
1                 86               88           81                 76   
2                 87               95           88                 89   

   skill_long_passing  skill_ball_control

In [92]:
df.tail()  # 별도의 숫자를 입력하지 않으면 5개가 기본

sofifa_id                                         player_url  \
18939     256679  https://sofifa.com/player/256679/kevin-angulo/...   
18940     257710  https://sofifa.com/player/257710/mengxuan-zhan...   
18941     250989  https://sofifa.com/player/250989/zhenghao-wang...   
18942     257697  https://sofifa.com/player/257697/zitong-chen/2...   
18943     257936   https://sofifa.com/player/257936/yue-song/210002   

           short_name       long_name  age         dob  height_cm  weight_kg  \
18939       K. Angulo    Kevin Angulo   24  1996-04-13        176         73   
18940  Zhang Mengxuan  Mengxuan Zhang   21  1999-04-26        177         70   
18941   Wang Zhenghao             王政豪   20  2000-06-28        185         74   
18942     Chen Zitong     Zitong Chen   23  1997-02-20        186         80   
18943        Song Yue        Yue Song   28  1991-11-20        185         79   

      nationality                            club_name  \
18939    Colombia                      América de Cali   
18940    China PR  Chongqing Dangdai Lifan FC SWM Team   
18941    China PR                      Tianjin TEDA FC   
18942    China PR        Shijiazhuang Ever Bright F.C.   
18943    China PR                      Tianjin TEDA FC   

                   league_name  league_rank  overall  potential  value_eur  \
18939  Colombian Liga Postobón          1.0       47         52      40000   
18940     Chinese Super League          1.0       47         52      35000   
18941     Chinese Super League          1.0       47         51      35000   
18942     Chinese Super League          1.0       47         51      40000   
18943     Chinese Super League          1.0       47         47      30000   

       wage_eur player_positions preferred_foot  international_reputation  \
18939       500               CM          Right                         1   
18940      1000               CB          Right                         1   
18941      1000               CB          Right                         1   
18942      1000               CM          Right                         1   
18943      2000               CM          Right                         1   

       weak_foot  skill_moves      work_rate body_type real_face  \
18939          2            2  Medium/Medium    Normal        No   
18940          2            2        Low/Low    Normal        No   
18941          2            2     Low/Medium    Normal        No   
18942          2            2        Low/Low      Lean        No   
18943          2            2        Low/Low      Lean        No   

       release_clause_eur player_tags team_position  team_jersey_number  \
18939             91000.0         NaN           SUB                28.0   
18940             57000.0         NaN           SUB                25.0   
18941             57000.0         NaN           RES                 2.0   
18942             65000.0         NaN           RDM                 4.0   
18943             47000.0         NaN           RES                33.0   

      loaned_from      joined  contract_valid_until nation_position  \
18939         NaN  2020-01-01                2024.0             NaN   
18940         NaN  2020-08-01                2020.0             NaN   
18941         NaN  2019-02-01                2022.0             NaN   
18942         NaN  2020-08-01                2023.0             NaN   
18943         NaN  2020-08-01                2022.0             NaN   

       nation_jersey_number  pace  shooting  passing  dribbling  defending  \
18939                   NaN  63.0      37.0     50.0       47.0       41.0   
18940                   NaN  58.0      23.0     26.0       27.0       50.0   
18941                   NaN  51.0      25.0     29.0       36.0       45.0   
18942                   NaN  61.0      35.0     48.0       48.0       44.0   
18943                   NaN  65.0      37.0     49.0       46.0       40.0   

       physic  gk_diving  gk_handling  gk_kicking  gk_reflexes

참고로, ```set_option```메서드를 통해 판다스의 여러가지 설정값을 바꿀 수 있습니다. 출력되는 열(칼럼) 수에는 제한을 두지 않는 편이지만, 행의 수는 큰 데이터셋을 다룰 때 메모리가 많이 소요되므로, ```display.max_rows``` 값은 건드리지 않는 편입니다. 저는 각각 None과 20을 적용해보겠습니다.

In [93]:
pd.set_option("display.max_columns", None)  # 칼럼 수 제한없음
pd.set_option("display.max_rows", 20) # 행의 수는 20개로 제한


칼럼은 총 몇 개일까요? ```len```함수를 통해 확인할 수 있습니다.

In [94]:
len(df.columns)

106

칼럼 수가 106개나 되는군요. 일반적으로 EDA를 진행하면서, 분석에 불필요한 행은 ```df.drop```메서드를 통해 솎아내는 편입니다. ```df.drop``` 메서드는 조금 뒤에 보여드리겠습니다.

우선 ```df.columns``` 커맨드를 통해 전체 칼럼은 어떤지, 어떤 데이터를 담고 있는지 한 번 확인해보겠습니다.

In [95]:
df.columns

Index(['sofifa_id', 'player_url', 'short_name', 'long_name', 'age', 'dob',
       'height_cm', 'weight_kg', 'nationality', 'club_name',
       ...
       'lwb', 'ldm', 'cdm', 'rdm', 'rwb', 'lb', 'lcb', 'cb', 'rcb', 'rb'],
      dtype='object', length=106)

# 이 데이터셋은?

축구 게임을 좋아하시는 분들이라면 전 세계의 이름있는 축구선수들의 정보(신상정보, 능력치나 연봉 등)가 들어있는 데이터프레임인 것을 바로 눈치채셨을 겁니다.

쭉 읽어보시면, 대부분의 칼럼명은 직관적으로 이해가 되지만, 그렇지 않은 칼럼들도 있을 것입니다. 일반적으로웹상에서 공유되는 이런 데이터셋들은 디스크립션 자료가 별도로 존재하는 편입니다. 우리가 불러온 데이터는 sofifa.com이라는 사이트에서 모든 현역 선수들의 오버롤 데이터를 크롤링한 자료입니다. 축구 팬이시라면 sofifa.com에서 여러분이 좋아하는 선수의 현재(또는 향후에 조정될) 스탯에 대해 검토하고 토론도 나눌 수 있답니다.

데이터셋 출처 : https://www.kaggle.com/datasets/stefanoleone992/fifa-21-complete-player-dataset?resource=download

축구를 별로 좋아하시지 않는 분들이라도 우리나라 사람이라면 손흥민 선수에 대한 애정과 관심은 클 거라고 생각합니다. 우리 데이터셋에 **손흥민** 선수가 들어있는지 인덱싱으로 한 번 확인해볼까요?

# 특정 행이나 특정 열을 선택하는 방법, 인덱싱(loc, iloc)

![](./h_son.jpg)

① 미리 알려드리면, 이 데이터프레임 안에서 손흥민 선수의 인덱스아이디는 71입니다(알고 있다고 가정). 이를 통해 손흥민 선수의 스탯을 확인하려면, 판다스 데이터프레임 메서드인 loc 또는 iloc를 활용하면 됩니다. ```df.loc[71]```을 입력해봅시다.

In [96]:
df.loc[71]  # 인덱스명으로 탐색

sofifa_id                                                200104
player_url    https://sofifa.com/player/200104/heung-min-son...
short_name                                               H. Son
long_name                                               손흥민 孙兴慜
age                                                          27
                                    ...                        
lb                                                         64+3
lcb                                                        54+3
cb                                                         54+3
rcb                                                        54+3
rb                                                         64+3
Name: 71, Length: 106, dtype: object

In [97]:
df.iloc[71]  # 인덱스 순서로 탐색

sofifa_id                                                200104
player_url    https://sofifa.com/player/200104/heung-min-son...
short_name                                               H. Son
long_name                                               손흥민 孙兴慜
age                                                          27
                                    ...                        
lb                                                         64+3
lcb                                                        54+3
cb                                                         54+3
rcb                                                        54+3
rb                                                         64+3
Name: 71, Length: 106, dtype: object

이 예제에서는 두 커맨드가 같은 결과를 출력했지만, 사실 iloc는 인덱스가 숫자가 아닌 경우에도 숫자로 행을 불러올 수 있는 메서드입니다. 예를 들어 데이터프레임의 인덱스가 "a", "b", "c", ... 인 경우에는 loc 메서드를 사용할 때는 ```df.loc["c"]```와 ```df.iloc[2]```는 동일한 행을 출력해줍니다.

그런데 우리는 사실 손흥민 선수의 인덱스(71)를 모르는 상태입니다. "short_name 칼럼의 문자열이 Son으로 끝나는 선수"라든지, "long_name 칼럼에 손흥민이라는 이름이 포함된 선수"라는 조건으로 검색을 해 보고 싶다면? 이런 경우를 위한 검색방법이 준비되어 있습니다. 아래 두 줄의 커맨드를 찬찬히 살펴보시기 바랍니다.

In [98]:
# short_name 칼럼의 값이 "Son"으로 끝나는 선수?
df[df.short_name.str.endswith("Son")]

sofifa_id                                         player_url short_name  \
71       200104  https://sofifa.com/player/200104/heung-min-son...     H. Son   
4267     252086  https://sofifa.com/player/252086/fco-javier-hi...        Son   

                           long_name  age         dob  height_cm  weight_kg  \
71                           손흥민 孙兴慜   27  1992-07-08        183         78   
4267  Francisco Javier Hidalgo Gómez   26  1994-03-30        178         72   

         nationality          club_name             league_name  league_rank  \
71    Korea Republic  Tottenham Hotspur  English Premier League          1.0   
4267           Spain         Levante UD  Spain Primera Division          1.0   

      overall  potential  value_eur  wage_eur player_positions preferred_foot  \
71         87         87   52500000    165000           LM, CF          Right   
4267       71         74    2600000     14000               RB          Right   

      international_reputation  weak_foot  skill_moves    work_rate  \
71                           3          5            4    High/High   
4267                         1          3            2  High/Medium   

                 body_type real_face  release_clause_eur  \
71    PLAYER_BODY_TYPE_366       Yes          99800000.0   
4267                  Lean        No           5700000.0   

                                            player_tags team_position  \
71    #Dribbler, #Engine, #Distance Shooter, #Acroba...            LM   
4267                                                NaN           SUB   

      team_jersey_number loaned_from      joined  contract_valid_until  \
71                   7.0         NaN  2015-08-28                2023.0   
4267                 2.0         NaN  2020-08-04                2024.0   

     nation_position  nation_jersey_number  pace  shooting  passing  \
71               NaN                   NaN  88.0      86.0     80.0   
4267             NaN                   NaN  74.0      46.0     61.0   

      dribbling  defending  physic  gk_diving  gk_handling  gk_kicking  \
71         87.0       43.0    69.0        NaN          NaN         NaN   
4267       67.0       67.0    68.0        NaN          NaN         NaN   

      gk_reflexes  gk_speed  gk_positioning  \
71            NaN       NaN             NaN   
4267          NaN       NaN             NaN   

                                          player_traits  attacking_crossing  \
71    Solid Player, Finesse Shot, Long Shot Taker (A...                  81   
4267                                                NaN                  65   

      attacking_finishing  attacking_heading_accuracy  \
71                     86                          68   
4267                   45                          63   

      attacking_short_passing  attacking_volleys  skill_dribbling  \
71                         83                 81               88   
4267                       67                 41               67   

      skill_curve  skill_fk_accuracy  skill_long_passing  skill_ball_control  \
71             85                 70                  70                  87   
4267           55                 45                  56                  66   

      movement_acceleration  movement_sprint_speed  movement_agility  \
71                       86                     90                86   
4267                     75                     74                67   

      movement_reactions  movement_balance  power_shot_power  power_jumping  \
71                    89                78                88             56   
4267                  68                68                44             64   

      power_stamina  power_strength  power_long_shots  mentality_aggression  \
71               87              64                90                    62   
4267             79              65                47                    63   

      mentality_interceptions  mentality_positioning  mentality_vision

다행히 우리 손흥민 선수를 찾기는 했지만, 스페인 프리메라리가의 하비에르 선수도 short_name 값이 Son이어서 같이 검색이 되었네요. 이번에는 long_name에 "손흥민"이 포함된 선수를 검색해보겠습니다.

In [99]:
df[df.long_name.str.contains("손흥민")]

sofifa_id                                         player_url short_name  \
71     200104  https://sofifa.com/player/200104/heung-min-son...     H. Son   

   long_name  age         dob  height_cm  weight_kg     nationality  \
71   손흥민 孙兴慜   27  1992-07-08        183         78  Korea Republic   

            club_name             league_name  league_rank  overall  \
71  Tottenham Hotspur  English Premier League          1.0       87   

    potential  value_eur  wage_eur player_positions preferred_foot  \
71         87   52500000    165000           LM, CF          Right   

    international_reputation  weak_foot  skill_moves  work_rate  \
71                         3          5            4  High/High   

               body_type real_face  release_clause_eur  \
71  PLAYER_BODY_TYPE_366       Yes          99800000.0   

                                          player_tags team_position  \
71  #Dribbler, #Engine, #Distance Shooter, #Acroba...            LM   

    team_jersey_number loaned_from      joined  contract_valid_until  \
71                 7.0         NaN  2015-08-28                2023.0   

   nation_position  nation_jersey_number  pace  shooting  passing  dribbling  \
71             NaN                   NaN  88.0      86.0     80.0       87.0   

    defending  physic  gk_diving  gk_handling  gk_kicking  gk_reflexes  \
71       43.0    69.0        NaN          NaN         NaN          NaN   

    gk_speed  gk_positioning  \
71       NaN             NaN   

                                        player_traits  attacking_crossing  \
71  Solid Player, Finesse Shot, Long Shot Taker (A...                  81   

    attacking_finishing  attacking_heading_accuracy  attacking_short_passing  \
71                   86                          68                       83   

    attacking_volleys  skill_dribbling  skill_curve  skill_fk_accuracy  \
71                 81               88           85                 70   

    skill_long_passing  skill_ball_control  movement_acceleration  \
71                  70                  87                     86   

    movement_sprint_speed  movement_agility  movement_reactions  \
71                     90                86                  89   

    movement_balance  power_shot_power  power_jumping  power_stamina  \
71                78                88             56             87   

    power_strength  power_long_shots  mentality_aggression  \
71              64                90                    62   

    mentality_interceptions  mentality_positioning  mentality_vision  \
71                       39                     88                83   

    mentality_penalties  mentality_composure  defending_marking  \
71                   73                   88                NaN   

    defending_standing_tackle  defending_sliding_tackle  goalkeeping_diving  \
71                         34                        33                  11   

    goalkeeping_handling  goalkeeping_kicking  goalkeeping_positioning  \
71                    13                   13                        6   

    goalkeeping_reflexes    ls    st    rs    lw    lf    cf    rf    rw  \
71                    10  84+3  84+3  84+3  86+0  87+0  87+0  87+0  86+0   

     lam   cam   ram    lm   lcm    cm   rcm    rm   lwb   ldm   cdm   rdm  \
71  86+1  86+1  86+1  85+2  79+3  79+3  79+3  85+2  68+3  63+3  63+3  63+3   

     rwb    lb   lcb    cb   rcb    rb  
71  68+3  64+3  54+3  54+3  54+3  64+3

저 검색문법이 이해가 안 되는 분들도 많을 것으로 생각합니다. ```df[]``` 대괄호 안의 커맨드인 ```df.long_name.str.contains("손흥민")```만 한 번 실행해보겠습니다.

In [100]:
df.long_name.str.contains("손흥민")

0        False
1        False
2        False
3        False
4        False
         ...  
18939    False
18940    False
18941    False
18942    False
18943    False
Name: long_name, Length: 18944, dtype: bool

부울 연산자(True, False)로만 이뤄진 18,944개의 칼럼입니다. 이 중 해당조건("손흥민"을 포함하는가)을 만족하는 행(71th)만 True를 리턴하고 있습니다. 이제 전체 커맨드를 보면, 아래와 같은 명령어라는 것을 이해하실 겁니다.

```df[[False, False, False, False, ... True, False, False, ...]]```

In [101]:
# 이해를 돕기 위한 작은 예를 보여드리겠습니다.
# 아래의 코드를 실행하면 True에 대응하는(세 번째 값인) "C"만 리턴됩니다.

df_temp = pd.DataFrame(["A", "B", "C", "D", "E"])
df_temp[[False, False, True, False, False]]

0
2  C

기본적으로 판다스 데이터프레임의 필터링이나 조건식 문법은 위와 같은 방식으로 내부적으로 부울 조건식을 활용합니다.

그럼 누구나 관심있는, 축구선수들의 연봉 이야기로 넘어가보겠습니다.
가장 연봉을 많이 받는 10명의 선수는 누구일까요?

![](l_messi.jpg)
![](c_ronaldo.jpg)
![](k_benzema.jpg)


가장 큰 값 n개를 리턴하는 메서드는 ```df.nlargest(n, 칼럼명)```입니다.

In [102]:
df.nlargest(10, "wage_eur")

sofifa_id                                         player_url  \
0      158023  https://sofifa.com/player/158023/lionel-messi/...   
7      192985  https://sofifa.com/player/192985/kevin-de-bruy...   
11     165153  https://sofifa.com/player/165153/karim-benzema...   
13     183277  https://sofifa.com/player/183277/eden-hazard/2...   
12     182521  https://sofifa.com/player/182521/toni-kroos/21...   
56     200145  https://sofifa.com/player/200145/carlos-henriq...   
9      153079  https://sofifa.com/player/153079/sergio-aguero...   
10     155862  https://sofifa.com/player/155862/sergio-ramos-...   
19     194765  https://sofifa.com/player/194765/antoine-griez...   
4      176580  https://sofifa.com/player/176580/luis-suarez/2...   

      short_name                          long_name  age         dob  \
0       L. Messi     Lionel Andrés Messi Cuccittini   33  1987-06-24   
7   K. De Bruyne                    Kevin De Bruyne   29  1991-06-28   
11    K. Benzema                      Karim Benzema   32  1987-12-19   
13     E. Hazard                        Eden Hazard   29  1991-01-07   
12      T. Kroos                         Toni Kroos   30  1990-01-04   
56      Casemiro  Carlos Henrique Venancio Casimiro   28  1992-02-23   
9      S. Agüero  Sergio Leonel Agüero del Castillo   32  1988-06-02   
10  Sergio Ramos                Sergio Ramos García   34  1986-03-30   
19  A. Griezmann                  Antoine Griezmann   29  1991-03-21   
4      L. Suárez           Luis Alberto Suárez Díaz   33  1987-01-24   

    height_cm  weight_kg nationality        club_name             league_name  \
0         170         72   Argentina     FC Barcelona  Spain Primera Division   
7         181         70     Belgium  Manchester City  English Premier League   
11        185         81      France      Real Madrid  Spain Primera Division   
13        175         74     Belgium      Real Madrid  Spain Primera Division   
12        183         76     Germany      Real Madrid  Spain Primera Division   
56        185         84      Brazil      Real Madrid  Spain Primera Division   
9         173         70   Argentina  Manchester City  English Premier League   
10        184         82       Spain      Real Madrid  Spain Primera Division   
19        176         73      France     FC Barcelona  Spain Primera Division   
4         182         86     Uruguay     FC Barcelona  Spain Primera Division   

    league_rank  overall  potential  value_eur  wage_eur player_positions  \
0           1.0       93         93   67500000    560000       RW, ST, CF   
7           1.0       91         91   87000000    370000          CAM, CM   
11          1.0       89         89   53000000    350000           CF, ST   
13          1.0       88         88   58000000    350000           LW, ST   
12          1.0       88         88   55000000    310000               CM   
56          1.0       89         89   59500000    310000              CDM   
9           1.0       89         89   53000000    300000               ST   
10          1.0       89         89   24500000    300000               CB   
19          1.0       87         87   50500000    290000       ST, CF, LW   
4           1.0       87         87   31500000    280000               ST   

   preferred_foot  international_reputation  weak_foot  skill_moves  \
0            Left                         5          4            4   
7           Right                         4          5            4   
11          Right                         4          4            4   
13          Right                         4          4            4   
12          Right                         4          5            3   
56          Right                         3          3            2   
9           Right                         4          4            4   
10          Right                         4          3            3   
19           Left                         4          3            4   
4           

아쉽지만, 손흥민 선수는 10위에 포함되지는 않네요.
그럼 손흥민 선수의 연봉은 몇 유로인지 확인해볼까요?

In [103]:
df[df.short_name == "H. Son"].wage_eur

71    165000
Name: wage_eur, dtype: int64

출력된 값은 165,000유로, 한화로 환산하면 약 2억2천만원인데요. 우리 연봉보다 조금밖에 많지 않아서 당황하셨나요? 사실 wage_eur은 연봉이 아니라 주급입니다. 52를 곱해보면 대략..

In [113]:
df[df.short_name == "H. Son"].wage_eur * 52 * 1342.64

71    1.151985e+10
Name: wage_eur, dtype: float64

이런 표기법이 익숙하지 않은 분도 계실 겁니다. e+6이면 백만원 e+10이면 백억원이니까 약 115억 정도가 되겠네요. 약간의 신태틱을 붙이면 조금 더 보기 편해지실 겁니다.

In [105]:
f'{int(df[df.short_name == "H. Son"].wage_eur * 52 * 1342.64):,}'

'11,519,851,200'

# 특정 행의 칼럼값 순위 확인(rank)

(사실 115억 정도면 굉장히 저평가된 것 같습니다. 2021~2022 프리미어리그 득점왕인데 말이죠...)
이 정도 금액이면 도대체 몇 위 정도일까요? ```df.rank()```로 알아보겠습니다.

In [106]:
df["rank_wage"] = df["wage_eur"].rank()
df[df.short_name == "H. Son"].rank_wage

71    18896.0
Name: rank_wage, dtype: float64

잉? 18896번째라고요? 뭔가 이상한 값이 나왔죠? 사실 rank() 메서드는 오름차순이 디폴트라 그렇습니다.
다시 계산해보겠습니다.

In [107]:
df["rank_wage"] = df["wage_eur"].rank(ascending=False)
df[df.short_name == "H. Son"].rank_wage

71    49.0
Name: rank_wage, dtype: float64

그래도 무려 프리미어리그 득점왕 손흥민 선수가 연봉 49위 밖에 아니라니 좀 아쉽기는 하네요...

재미로 몇 가지 다른 값도 한 번 탐색해보겠습니다.
축구선수 중 키가 큰 선수들은 대부분 어떤 나라 사람들일까요?
우선 최고로 키가 큰 선수의 신장을 한 번 알아보겠습니다.

In [108]:
# df["height_cm"].max() 와 동일
df.height_cm.max()

206

# 정렬(sort_values) 사용해보기

최장신 선수는 무려 206센티미터네요. 그럼 키가 2미터 이상인 선수들 열 명만 내림차순으로 출력해보겠습니다. 정렬은 ```df.sort_values``` 메서드를 사용하면 됩니다.

In [109]:
df[["short_name", "nationality", "league_name", "height_cm", "league_rank"]].sort_values("height_cm", ascending=False)[:10]

short_name     nationality                   league_name  \
10190              T. Holý  Czech Republic            English League One   
11374                A. Ba      Mauritania                French Ligue 2   
4449         C. Pantilimon         Romania             Turkish Süper Lig   
7381           K. Scherpen     Netherlands            Holland Eredivisie   
11578        S. Maierhofer         Austria  Austrian Football Bundesliga   
6643   V. Milinković-Savić          Serbia               Italian Serie A   
18199             D. Thiam         Senegal               Italian Serie B   
17713          D. Turnbull   United States            English League One   
18658     K. McKenzie-Lyle          Guyana            English League Two   
9768            T. Rönning          Sweden           Swedish Allsvenskan   

       height_cm  league_rank  
10190        206          3.0  
11374        203          2.0  
4449         203          1.0  
7381         202          1.0  
11578        202          1.0  
6643         202          1.0  
18199        202          2.0  
17713        201          3.0  
18658        201          4.0  
9768         201          1.0

In [110]:
# 위 명령어와 아래 명령어는 동일한 결과를 반환합니다.
df.sort_values("height_cm", ascending=False)[["short_name", "nationality", "league_name", "height_cm", "league_rank"]][0:10]

short_name     nationality                   league_name  \
10190              T. Holý  Czech Republic            English League One   
11374                A. Ba      Mauritania                French Ligue 2   
4449         C. Pantilimon         Romania             Turkish Süper Lig   
7381           K. Scherpen     Netherlands            Holland Eredivisie   
11578        S. Maierhofer         Austria  Austrian Football Bundesliga   
6643   V. Milinković-Savić          Serbia               Italian Serie A   
18199             D. Thiam         Senegal               Italian Serie B   
17713          D. Turnbull   United States            English League One   
18658     K. McKenzie-Lyle          Guyana            English League Two   
9768            T. Rönning          Sweden           Swedish Allsvenskan   

       height_cm  league_rank  
10190        206          3.0  
11374        203          2.0  
4449         203          1.0  
7381         202          1.0  
11578        202          1.0  
6643         202          1.0  
18199        202          2.0  
17713        201          3.0  
18658        201          4.0  
9768         201          1.0

# 다중 정렬(sort_values().sort_values()...)

sort_values 메서드를 여러 번 사용해서 다중 정렬도 구현할 수 있습니다.
1차로 키 오름차순, 2차로 나이 내림차순으로 검색해보려면 sort_values를 연달아 두 번 사용하면 됩니다(일명 체이닝).

이 때 1차를 먼저 입력해야 할까요, 2차를 먼저 입력해야 할까요?
가만 생각해보면 먼저 입력하는 칼럼이 후순위가 됩니다. 뒤에서 섞이니까요.

In [111]:
df.sort_values("age", ascending=False).sort_values("height_cm")[["short_name", "nationality", "age", "height_cm"]][0:10]

short_name nationality  age  height_cm
12541    H. Nakagawa       Japan   25        155
5622      N. Barrios   Argentina   22        156
4494        J. Plata     Ecuador   28        157
16323      J. Nisbet   Australia   21        158
3611   C. Songkrasin    Thailand   26        158
880       M. Moralez   Argentina   33        158
16694      J. Romero   Argentina   19        159
7550    M. Domínguez      Mexico   20        159
7255      E. Oztumer     England   29        160
4551    V. Hernández    Colombia   31        160

In [ ]:
df.query("shooting > passing")

In [ ]:
df.iloc[:10, 5:10]

In [ ]:
df.loc[3]

In [ ]:
df.loc[[0, 1, 3], ["short_name", "nationality"]]

In [ ]:
df.height_cm.dtypes

In [ ]:
df.info()

In [ ]:
df.dtypes

In [ ]:
df.sample(n=200)

In [ ]:
df.sample(frac=0.25)

In [ ]:
df.skill_moves.unique()

In [ ]:
df.nationality.nunique()

In [ ]:
df.nunique()

In [ ]:
df["rank_calc"] = df["value_eur"].rank()

In [ ]:
df.head(3)

In [ ]:
df[df.nationality.str.contains("Korea")].head()

In [ ]:
df[df.nationality == "Korea Republic"]

In [ ]:
df.rename(columns={"weight_kg": "Weight(kg)", "height_cm": "Height(cm)"})

In [ ]:
df.height_cm.head()  # 바뀌지 않음

In [ ]:
df.pace.head()

In [ ]:
df.pace.fillna(0, inplace=True)

In [ ]:
df.pace.head()

In [ ]:
df.pace.fillna(df["pace"].mean(), inplace=True)

In [ ]:
df.groupby("nationality")["value_eur"].sum()

In [ ]:
df.groupby(['nationality', 'league_rank'])[['value_eur', 'wage_eur']].agg([np.mean, np.median])

In [ ]:
df['league_rank'].value_counts(ascending=True)

In [ ]:
pd.crosstab(df['league_rank'], df['international_reputation'])